### ISSUE TITLE
Windows 10, Chrome - Granting Access to Files (System Files Error)

### ISSUE THREAD

**Issue by BrookfieldIT** [photopea/photopea#5261]
Hello,

We're currently using Photopea for our iMedia course in school.
We've noticed however that a few students have been having some issues when trying to access exisiting PSD files, saving PSD files or even trying to grant PhotoPea access to a folder.

You have to grant PhotoPea access to a specific folder but it's showing up an error of:

![image](https://user-images.githubusercontent.com/122614091/212325034-b6e206bc-1a8e-4c92-b3bb-2b91a07459ea.png)

The folder is brand new and there is nothing in it (I had view Hidden files on as well):

![image](https://user-images.githubusercontent.com/122614091/212325202-6219a2e3-c377-4fa6-b2ef-9d0fd4bc5bb0.png)

Any help would be greatly appreciated.

Kind Regards,
Luke

**Comment by photopea** [comment]
This is between your web browser and your operating system. I use Chrome on Windows 10, and it never happened to me.

Maybe try a different folder, or check for hidden files in your folder.

**Comment by BrookfieldIT** [comment]
Hi PhotoPea,

Thank you for your quick reply, I'm not sure this is the case, I have been able to replicate granting access to a completely different folder and then trying it again with the original folder works and vice versa.

Could you provide the logic behind what it checks when selecting the file or explain what exactly you're specifying as a system file? .INI ect.

Perhaps there is way to ask PhotoPea to simply ignore/not show these alleged system files?

Thank you for your help.

Kind Regards,
Luke

**Comment by photopea** [comment]
As I said, it is a web browser who does not allow a website to access a folder. I dont know what the browser logic is. We can not do anything about it. 

Many other websites have a similar issue: https://github.com/microsoft/vscode/issues/144753
@tomayac do you know anything about it?

**Comment by tomayac** [comment]
As highlighted in the VS Code issue, the complete blocklist per platform is in this [code snippet](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/file_system_access/chrome_file_system_access_permission_context.cc;l=172-235;drc=359c4471ef9ffdf94d1080815d51f28ce0302b94). You mention this is in an education environment? Maybe the user directories are somehow protected? I can't reproduce with your exact folder structure.

**Comment by photopea** [comment]
I just noticed it. When I go to www.Photopea.com and choose "This Device" on the left, I can not give Photopea access to C:/Users/ivank/Desktop, same for /Documents and /Downloads  :( I really think this is wrong.

**Comment by BrookfieldIT** [comment]
^ This is the exact issue we're having.

**Comment by tomayac** [comment]
Regular apps can trigger a download without user action (via `<a download="supercookie.txt">`). This download lands in `Downloads`. If you then grant access to the entire `Downloads` folder, sites can read and write `supercookie.txt` and abuse this file as a super cookie. That's why this folder (but not subfolders of it) is blocked.

**Comment by photopea** [comment]
The browser clearly says, that a webap will have a full access to that folder, and can delete anything, etc.

What do you mean by "regular apps triggering a download"? A website can start a download only with a user action. Why would non-web apps use the Download folder?

**Comment by tomayac** [comment]
Just leaving this here: https://sun-panoramic-mice.glitch.me/.

```js
const file = new File(['Hi there'], {type: 'text/plain'});
const a = document.createElement('a');
a.download = 'yolo.txt';
a.href = URL.createObjectURL(file);
a.click();
```

In Chrome, this downloads by just browsing the site.

**Comment by photopea** [comment]
So the problem is, that a website saved a file without a user noticing, or another website rewrote that file without the user realizing it?

**Comment by tomayac** [comment]
Download without user interaction is a bad inheritance from the past. The problem is that this behavior can be abused if an app has access to the downloads folder and cross-site tracking is possible (when different sites obtain access to the same file).

**Comment by photopea** [comment]
In my case, opening this https://sun-panoramic-mice.glitch.me/ asks me to confirm the download ofa a file (I can refuse it). If a browser saves a file without telling the user about it (not sure if any browser does that, but you say they do), it is a bug in that browser. So you are limiting the user because of the incapability to solve a bug.

If two websites wanted to share the information about who has visited them, they would simply exchange a list of IP addresses of their visitors with each other. There is no need to do such crazy stuff as asking the user to give access to a device folder.

**Comment by Grimthorr** [comment]
> As highlighted in the VS Code issue, the complete blocklist per platform is in this [code snippet](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/file_system_access/chrome_file_system_access_permission_context.cc;l=172-235;drc=359c4471ef9ffdf94d1080815d51f28ce0302b94). You mention this is in an education environment? Maybe the user directories are somehow protected? I can't reproduce with your exact folder structure.

This breaking change does make sense from a security standpoint but is typically at the cost of convenience and user experience. Is there a flag in Chrome to override this behaviour?

Additionally, more recent versions of Chrome are now also [blocking UNC paths](https://source.chromium.org/chromium/chromium/src/+/main:chrome/browser/file_system_access/chrome_file_system_access_permission_context.cc;l=1472;bpv=1;bpt=0;drc=915abdbb04e2d71e288a87786a5293a83080db37;dlc=7198a5329306ec57925e9a719376746c8209b732), which seems to affecting OP; they are likely employing [Folder Redirection](https://learn.microsoft.com/en-us/windows-server/storage/folder-redirection/folder-redirection-rup-overview), which means the entirely of the user's redirected home folders (including subfolders) are now considered sensitive. This seems to have been introduced in builds `109.x` of Chrome; UNC paths were allowed in `108.0.5359.126`, but are disallowed in `109.0.5414.75`.

Worryingly, a user's `Pictures` folder is not blocked, while this is arguably more sensitive than the `Downloads` folder. This could be used as a workaround for others, but not for OP where this is still redirected to a UNC path.

In summary, these changes in Chrome effectively: 
- Force standard users to use a subfolder within their home folders, or use a location outside this structure.
- Break the use of redirected home folders, meaning these user's must use an unusual location (outside of their home folders).

**Comment by photopea** [comment]
@tomayac @Grimthorr I am still very confused by this attitude of browser makers. I thought that the goal of browsers was to allow websites to be as powerful as native apps.

There are people who download and run an EXE file from a company they have never heard of, and that EXE can access or delete any file on a hard drive. Then, there are people who have been using a web app for months or years, or even know the author of that web app in person, but they still can not let the web app access their Downloads folder.

**Comment by tomayac** [comment]
If you disagree with the [assessment in the spec](https://wicg.github.io/file-system-access/#privacy-wide-access), feel free to [file a spec issue](https://github.com/WICG/file-system-access/issues).

If, given the current spec, you disagree with Chromium's implementation, feel free to [file a new Chromium bug](https://crbug.com/new). (Feel free to paste the bug link here, so I can help triage.)

**Comment by jimmywarting** [comment]
fyi, if you drag and drop in a folder in to your web app, then you don't have the same restrictions and you can basically choose any folder you like.

When dropping a folder call `item.getAsFileSystemHandle()`

**Comment by tomayac** [comment]
FWIW, the drag and drop loophole for blocklisted directories to be accessible will be closed soon. Do not rely on it in the future.

**Comment by prasiddharaj** [comment]
Hi I am facing the same issue on Chrome/Edge version 109. Any resolutions yet?

**Comment by photopea** [comment]
I have stopped using the file-system-access API. It should work now. Could you check, if opening files works? And if you can save them?

**Comment by d-mouse** [comment]
![image](https://user-images.githubusercontent.com/1980474/214473837-b80d5298-e0af-410b-a8e9-f39a6e90ba4f.png)

We have the same isssue in Chrome Version 109.0.5414.120 (Official Build) (32-bit) and Edge 

Can create a new project and save it okay but then try reopen the same file from File --> Open, and we get the error.

Drag and drop the same file from explorer onto the web app works fine so that is the workaround.

**Comment by photopea** [comment]
So, Chrome team seems to be working on it, let's hope they will fix it as soon as possible.

https://bugs.chromium.org/p/chromium/issues/detail?id=1408321

**Comment by camoody1** [comment]
Nice

**Comment by jimmywarting** [comment]
> https://bugs.chromium.org/p/chromium/issues/detail?id=1408321

star'ed it. (not b/c i use your service but b/c i'm using file system access myself on my website)
